src/io_darwin.zig

```diff
@@ -0,0 +1,691 @@
+const std = @import("std");
+const os = std.os;
+const assert = std.debug.assert;
+
+const FIFO = @import("fifo.zig").FIFO;
+const Time = @import("time.zig").Time;
+const buffer_limit = @import("io.zig").buffer_limit;
+
+/// Logging that was used for debugging purposes. Can safely be deleted
+fn log(comptime fmt: []const u8, args: anytype) void {
+    // if (@hasDecl(@import("root"), "is_main")) return;
+    // if (args[1] == .timeout) return;
+    // std.log.info(fmt, args);
+}
+
+pub const IO = struct {
+    kq: os.fd_t,
+    time: Time = .{},
+    timeouts: FIFO(Completion) = .{},
+    completed: FIFO(Completion) = .{},
+    io_pending: FIFO(Completion) = .{},
+
+    pub fn init(entries: u12, flags: u32) !IO {
+        return IO{ .kq = try os.kqueue() };
+    }
+
+    pub fn deinit(self: *IO) void {
+        os.close(self.kq);
+    }
+
+    /// Pass all queued submissions to the kernel and peek for completions.
+    pub fn tick(self: *IO) !void {
+        return self.flush(false);
+    }
+
+    /// Pass all queued submissions to the kernel and run for `nanoseconds`.
+    /// The `nanoseconds` argument is a u63 to allow coercion to the i64 used
+    /// in the __kernel_timespec struct.
+    pub fn run_for_ns(self: *IO, nanoseconds: u63) !void {
+        var timed_out = false;
+        var completion: Completion = undefined;
+        const on_timeout = struct {
+            fn callback(timed_out_ptr: *bool, _completion: *Completion, _result: TimeoutError!void) void {
+                timed_out_ptr.* = true;
+            }
+        }.callback;
+
+        // Submit a timeout which sets the timed_out ptr value to true to terminate the loop 
+        const timed_out_ptr = &timed_out;
```

Comment:
I would like to understand better why we capture the pointer here and then use that in the callback wrapper etc. I guess it must be for some kind of pointer stability?

Reply:
How does this interact with reentrancy and recursion?

Reply:
It's effectively doing `while (!timeout) try self.flush(true)`. The `timeout` bool is changed by the nanoseconds Completion expiring. For the completion to change the `timeout` local variable, it needs its address since its callback is executed inside `self.flush()`. 

The while loop also reads from the `timeout` pointer instead of the variable to further hint at LLVM that the variable changes externally to the function, and that it should not cache its result (i.e. `const t = timeout; while (t) ...`). We could make this a guarantee by doing `while (!@ptrCast(*volatile bool, timeout_ptr).*)` but idk if that was overkill. Should I document this in the code?

As for recursion, the loop only stops when that specific Completion's timeout has expired so it should handle that correctly (unless i'm mistaken by the prompt).

Reply:
Cool, would it not be a compiler bug if LLVM decided to cache the `timeout` result without regard to external changes? If so, then I think we don't need the `volatile` hammer, but documenting this in the code as a comment would be great.

Reply:
You may be right. I switched it to use the timeout variable directly and added a comment.